pkg/ovirt/actuator.go

```diff
@@ -0,0 +1,294 @@
+	return nil
+}
+
+func (a *OvirtActuator) syncCredentialSecret(cr *minterv1.CredentialsRequest, ovirtCreds *OvirtCreds, existingSecret *corev1.Secret, userPolicy string, logger log.FieldLogger) error {
+	sLog := logger.WithFields(log.Fields{
+		"targetSecret": fmt.Sprintf("%s/%s", cr.Spec.SecretRef.Namespace, cr.Spec.SecretRef.Name),
+		"cr":           fmt.Sprintf("%s/%s", cr.Namespace, cr.Name),
+	})
+
+	if existingSecret == nil || existingSecret.Name == "" {
+		if ovirtCreds == nil {
+			msg := "new access key secret needed but no key data provided"
+			sLog.Error(msg)
+			return &actuatoriface.ActuatorError{
+				ErrReason: minterv1.CredentialsProvisionFailure,
+				Message:   msg,
+			}
+		}
+		sLog.Info("creating secret")
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      cr.Spec.SecretRef.Name,
+				Namespace: cr.Spec.SecretRef.Namespace,
+				Annotations: map[string]string{
+					minterv1.AnnotationCredentialsRequest: fmt.Sprintf("%s/%s", cr.Namespace, cr.Name),
+				},
+			},
+			Data: map[string][]byte{
+				urlKey:      []byte(ovirtCreds.URL),
+				usernameKey: []byte(ovirtCreds.Username),
+				passwordKey: []byte(ovirtCreds.Passord),
+				cafileKey:   []byte(ovirtCreds.CAFile),
+			},
+		}
+
+		err := a.Client.Create(context.TODO(), secret)
+		if err != nil {
+			sLog.WithError(err).Error("error creating secret")
+			return err
+		}
+		sLog.Info("secret created successfully")
+		return nil
+	}
+
+	// Update the existing secret:
+	sLog.Debug("updating secret")
+	origSecret := existingSecret.DeepCopy()
+	if existingSecret.Annotations == nil {
+		existingSecret.Annotations = map[string]string{}
+	}
+	existingSecret.Annotations[minterv1.AnnotationCredentialsRequest] = fmt.Sprintf("%s/%s", cr.Namespace, cr.Name)
+	if ovirtCreds != nil {
+		existingSecret.Data = map[string][]byte{
+			urlKey:      []byte(ovirtCreds.URL),
+			usernameKey: []byte(ovirtCreds.Username),
+			passwordKey: []byte(ovirtCreds.Passord),
+			cafileKey:   []byte(ovirtCreds.CAFile),
+		}
+	}
+
+	if !reflect.DeepEqual(existingSecret, origSecret) {
+		sLog.Info("target secret has changed, updating")
+		err := a.Client.Update(context.TODO(), existingSecret)
+		if err != nil {
+			msg := "error updating secret"
+			sLog.WithError(err).Error(msg)
+			return &actuatoriface.ActuatorError{
+				ErrReason: minterv1.CredentialsProvisionFailure,
+				Message:   msg,
+			}
+		}
+	} else {
+		sLog.Debug("target secret unchanged")
+	}
+
+	return nil
+}
+
+func (a *OvirtActuator) loadExistingSecret(cr *minterv1.CredentialsRequest) (*corev1.Secret, error) {
+	logger := a.getLogger(cr)
+
+	// Check if the credentials secret exists, if not we need to inform the syncer to generate a new one:
+	existingSecret := &corev1.Secret{}
+	err := a.Client.Get(context.TODO(), types.NamespacedName{Namespace: cr.Spec.SecretRef.Namespace, Name: cr.Spec.SecretRef.Name}, existingSecret)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Debug("secret does not exist")
```

Comment:
I want to create a new secret. I can use the err type NotFound for flow controll in the caller, right now I use the fact that the secret it empty. The handling is more generic like that I think.

Reply:
help me understand the thinking when passing around a non-nil (but otherwise empty) secret.

i see in sync() we loadExistingSecret() and it can return a non-nil (but empty) secret if the secret isn't found.

then later in syncCredentialsSecret() it checks for both `existingSecret == nil || .Name == ""`. to me this could be simplified to just a check for `existingSecret == nil` if loadExistingSecret() returned a nil, nil. that way `existingSecret == nil` is the signal that there is no existing secret.

in summary, syncCredentialsSecret() doesn't use anything inside of existingSecret when it is empty, so why return the non-nil secret if syncCredentialsSecret() goes and creates a brand new secret object when there is no existing secret?

Reply:
Updating to a simpler version. PTAL

Reply:
sgtm